### PR TITLE
Update indexing entries when deleting a folder

### PR DESCRIPTION
**Pull request by sixstone-qq**
* It honors the HARD_DELETE flag as described.

Another things done in this PR:

* libmapi: Give support to not send any flag
* Delete mapistore first on mapistore root folders
    * As this honors when it is unable to delete due to flags
* Added new mapitest to test DeleteFolder ROP flags
    * Tested against OpenChange and Exchange 2010

### CODE REVIEW

**Comment by jkerihuel** [comment]
afaics LGTM

**Comment by jkerihuel** [comment]
:hand: reviewing one stuff related to that

**Comment by jkerihuel** [comment]
ok, one concern I have is that it introduces some inconsistencies in the semantics of the API. Basically we do the indexing_delete, but we don't do the indexing_add at create_folder layer. In addition to this, this would be the only case where indexing_record_* API is used which makes me wonder: what about messages?

I think a f2f discussion would probably easier to work this out.
@blaxter, @sixstone-qq: can we do it today?

**Comment by sixstone-qq** [comment]
@jkerihuel I moved the indexing management to `emsmdbp_object` layer and fixes other things along the way as stated in the PR description.

File: mapiproxy/libmapistore/mapistore_interface.c [hunk 1]
```diff
@@ -567,10 +571,15 @@ _PUBLIC_ enum mapistore_error mapistore_folder_delete(struct mapistore_context *
 	}
 
 	/* Step 2a. Handle deletion of normal messages */
-	ret = mapistore_folder_get_child_fmids(mstore_ctx, context_id, folder, MAPISTORE_MESSAGE_TABLE, mem_ctx, &child_fmids, &child_count);
+	ret = mapistore_folder_get_child_fmids(mstore_ctx, context_id, folder, MAPISTORE_MESSAGE_TABLE, local_mem_ctx, &child_fmids, &child_count);
 	if (ret != MAPISTORE_SUCCESS) {
 		goto end;
 	}
+
+	deleted_fmids = talloc_zero_array(mem_ctx, uint64_t, 1);
```
Comment (jkerihuel) [inline mapiproxy/libmapistore/mapistore_interface.c]: OOM check missing

File: mapiproxy/libmapistore/mapistore_interface.c [hunk 2]
```diff
@@ -578,6 +587,12 @@ _PUBLIC_ enum mapistore_error mapistore_folder_delete(struct mapistore_context *
 				if (ret != MAPISTORE_SUCCESS) {
 					goto end;
 				}
+
+				*current_deleted_fmid = child_fmids[i];
+				deleted_count++;
+				deleted_fmids = talloc_realloc(mem_ctx, deleted_fmids, uint64_t,
```
Comment (jkerihuel) [inline mapiproxy/libmapistore/mapistore_interface.c]: OOm check missing

File: mapiproxy/libmapistore/mapistore_interface.c [hunk 3]
```diff
@@ -598,6 +613,12 @@ _PUBLIC_ enum mapistore_error mapistore_folder_delete(struct mapistore_context *
...
```
Comment (jkerihuel) [inline mapiproxy/libmapistore/mapistore_interface.c]: OOM check missing

File: mapiproxy/libmapistore/mapistore_interface.c [hunk 4]
```diff
@@ -623,6 +644,12 @@ _PUBLIC_ enum mapistore_error mapistore_folder_delete(struct mapistore_context *
...
```
Comment (jkerihuel) [inline mapiproxy/libmapistore/mapistore_interface.c]: OOM check missing

File: mapiproxy/servers/default/emsmdb/emsmdbp_object.c [hunk 5]
```diff
@@ -1445,15 +1445,62 @@ _PUBLIC_ enum mapistore_error emsmdbp_folder_move_folder(struct emsmdbp_context
...
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error.
+*/
+_PUBLIC_ enum mapistore_error emsmdbp_folder_delete_indexing_records(struct mapistore_context *mstore_ctx, uint32_t context_id, char *username, uint64_t fid, uint64_t *deleted_fmids, uint32_t deleted_fmids_count, uint8_t flags)
+{
+        enum mapistore_error    ret;
+        uint8_t                 delete_type_flag;
+        uint32_t                i;
+
+        delete_type_flag = (flags & DELETE_HARD_DELETE) ? MAPISTORE_PERMANENT_DELETE : MAPISTORE_SOFT_DELETE;
+        ret = mapistore_indexing_record_del_fid(mstore_ctx, context_id, username, fid, delete_type_flag);
```
Comment (jkerihuel) [inline mapiproxy/servers/default/emsmdb/emsmdbp_object.c]: missing retval check

File: mapiproxy/servers/default/emsmdb/emsmdbp_object.c [hunk 6]
```diff
@@ -1445,15 +1445,62 @@ _PUBLIC_ enum mapistore_error emsmdbp_folder_move_folder(struct emsmdbp_context
...
+   -# DEL_MESSAGES Delete all the messages in the folder
+   -# DEL_FOLDERS Delete the subfolder and all of its subfolders
+   -# DELETE_HARD_DELETE Hard delete the folder
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error.
+*/
 _PUBLIC_ enum mapistore_error emsmdbp_folder_delete(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *parent_folder, uint64_t fid, uint8_t flags)
 {
 	enum mapistore_error	ret;
 	enum MAPISTATUS		mapiret;
```
Comment (jkerihuel) [inline mapiproxy/servers/default/emsmdb/emsmdbp_object.c]: can we use retval instead of mapiret :pray:

**Comment by blaxter** [comment]
:+1: working like a charm

**Comment by jkerihuel** [comment]
:+1:

**Pull request merged**